### PR TITLE
RNN python support and a text classification example

### PR DESCRIPTION
**Pull request by psyyz10**
## What changes were proposed in this pull request?

Wrap rnns with python support. And example illustrated as a text classification example

## How was this patch tested?

Python unit test

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/611

### CODE REVIEW

File: dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala [hunk 1]
```diff
@@ -145,6 +145,46 @@ class PythonBigDL[T: ClassTag](implicit ev: TensorNumeric[T]) extends Serializab
...
+    inputSize: Int,
+    hiddenSize: Int,
+    activation: TensorModule[T]): RnnCell[T] = {
+    RnnCell[T](inputSize, hiddenSize, activation)
+  }
+
+  def createTimeDistributedCriterion(
+    critrn: TensorCriterion[T],
+    sizeAverage: Boolean): TimeDistributedCriterion[T] = {
+    TimeDistributedCriterion[T](critrn)
```
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala]: TimeDistributedCriterion[T](critrn, sizeAverage)

File: dl/src/main/python/nn/layer.py [hunk 2]
```diff
@@ -214,16 +214,81 @@ def __init__(self, kw,
...
+    creating: createTanh
+    creating: createRNNCell
+    '''
+
+    def __init__(self,
+                 input_size,
+                 hidden_size,
+                 activation,
+                 bigdl_type="float"):
+        super(RNNCell, self).__init__(None, bigdl_type, input_size, hidden_size, activation)
```
Comment (jason-dai) [inline dl/src/main/python/nn/layer.py]: activation.value?

File: dl/src/main/python/nn/layer.py [hunk 3]
```diff
@@ -214,16 +214,81 @@ def __init__(self, kw,
...
+
+class TimeDistributed(Model):
+    '''
+    >>> td = TimeDistributed(Linear(2, 3))
+    creating: createLinear
+    creating: createTimeDistributed
+    '''
+
+    def __init__(self, model, bigdl_type="float"):
+        super(TimeDistributed, self).__init__(None, bigdl_type, model)
```
Comment (jason-dai) [inline dl/src/main/python/nn/layer.py]: model.value?

File: dl/src/main/python/nn/layer.py [hunk 4]
```diff
@@ -214,16 +214,81 @@ def __init__(self, kw,
...
+
+class TimeDistributedCriterion(Model):
```
Comment (jason-dai) [inline dl/src/main/python/nn/layer.py]: Criterion is not a model

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/LSTM.scala [hunk 5]
```diff
@@ -36,10 +36,10 @@ import scala.reflect.ClassTag
 @SerialVersionUID(- 8176191554025511686L)
 class LSTM[T : ClassTag] (
   val inputSize: Int,
-  val hiddenSize: Int)
+  val hiddenSize: Int,
+  val p: Double = 0)
```
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/LSTM.scala]: scaladoc for p

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/GRU.scala [hunk 6]
```diff
@@ -33,10 +33,10 @@ import scala.reflect.ClassTag
 @SerialVersionUID(6717988395573528459L)
 class GRU[T : ClassTag] (
   val inputSize: Int,
-  val outputSize: Int)
+  val outputSize: Int,
+  val p: Double = 0.0)
```
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/GRU.scala]: scaladoc for p

File: dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala (hunk 1 above)
Comment (psyyz10) [inline dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala]: Done

File: dl/src/main/python/nn/layer.py (hunk 4 above)
Comment (psyyz10) [inline dl/src/main/python/nn/layer.py]: Done

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/LSTM.scala (hunk 5 above)
Comment (psyyz10) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/LSTM.scala]: Removed

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/GRU.scala (hunk 6 above)
Comment (psyyz10) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/GRU.scala]: Removed P

File: dl/src/main/python/models/textclassifier/textclassifier.py [hunk 7]
```diff
@@ -61,21 +61,36 @@ def to_sample(vectors, label, embedding_dim):
     # flatten nested list
     flatten_features = list(itertools.chain(*vectors))
     features = np.array(flatten_features, dtype='float').reshape(
-        [sequence_len, embedding_dim]).transpose(1, 0)
+        [sequence_len, embedding_dim])
+
+    if model_type == "cnn" or model_type == "CNN":
```
Comment (zhichao-li) [inline dl/src/main/python/models/textclassifier/textclassifier.py]: model_type.lower() == "cnn" ?

File: dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala [hunk 8]
```diff
@@ -144,6 +144,44 @@ class PythonBigDL[T: ClassTag](implicit ev: TensorNumeric[T]) extends Serializab
...
+    inputSize: Int,
+    hiddenSize: Int): LSTM[T] = {
+    LSTM[T](inputSize, hiddenSize)
+  }
+
+  def createRecurrent(): Recurrent[T] = {
+    Recurrent[T]()
+  }
+
+  def createTimeDistributedCriterion(
```
Comment (zhichao-li) [inline dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala]: move this to criterion.py

**Comment by zhichao-li** [comment]
LTGM except for minor changes.

**Pull request merged**